### PR TITLE
HDFStore: Fix empty result of keys() method on non-pandas hdf5 file

### LINKED ISSUE #29916: HDF5: empty groups and keys
Hi,

With some of the hdf5 files I have, `pandas.HDFStore.groups()` returns an empty list. (as does `.keys()` which iterates over the groups). However, the data are accessible via `.get()` or `.get_node()`.

This is related to #21543 and #21372 where the `.groups()` logic was changed, in particular using `self._handle.walk_groups()` instead of `self._handle.walk_nodes()`, now to be found here:
https://github.com/pandas-dev/pandas/blob/ea2e26ae7d700d7fd363ea5bfc05d2fe3fb8a5ee/pandas/io/pytables.py#L1212

#### Current Output

```python
>>> hdf.groups()
[]
```
```python
>>> hdf.keys()
[]
```

#### Expected Ouptut

List of groups and keys as visible with e.g. `h5dump`.
**Note:** Changing the aforementioned line back to use `.walk_nodes()` fixes the issue and lists the groups and keys properly:

```python
>>> hdf.groups()
[/Data/Table Layout (Table(69462,), zlib(4)) ''
   description := {
...
/Data/Array Layout/2D Parameters/Data Parameters (Table(15,)) ''
   description := {
   "mnemonic": StringCol(itemsize=8, shape=(), dflt=b'', pos=0),
   "description": StringCol(itemsize=48, shape=(), dflt=b'', pos=1),
   "isError": Int64Col(shape=(), dflt=0, pos=2),
   "units": StringCol(itemsize=7, shape=(), dflt=b'', pos=3),
   "category": StringCol(itemsize=31, shape=(), dflt=b'', pos=4)}
   byteorder := 'little'
   chunkshape := (642,)]]
```
```python
>>> hdf.keys()
['/Data/Table Layout',
 '/Metadata/Data Parameters',
 '/Metadata/Experiment Notes',
 '/Metadata/Experiment Parameters',
 '/Metadata/Independent Spatial Parameters',
 '/Metadata/_record_layout',
 '/Data/Array Layout/Layout Description',
 '/Data/Array Layout/1D Parameters/Data Parameters',
 '/Data/Array Layout/2D Parameters/Data Parameters']
```

#### Fix

One solution would be (I guess) to revert #21543, another to fix at least `.keys()` to use `._handle.walk_nodes()` instead of `.groups()` in
https://github.com/pandas-dev/pandas/blob/ea2e26ae7d700d7fd363ea5bfc05d2fe3fb8a5ee/pandas/io/pytables.py#L562

Could also be that it is a bug in `pytables`.

#### Problem background

I was trying to figure out why some hdf5 files open fine with `pandas` but fail with `dask`.
The reason is that `dask` allows wildcards and iterates over the keys to find valid ones. If `.keys()` is empty, reading the files with `dask` fails.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.3.final.0
python-bits      : 64
OS               : Linux
OS-release       : 3.10.0-957.27.2.el7.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : C
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 19.3.1
setuptools       : 42.0.1.post20191125
Cython           : None
pytest           : 5.0.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.2
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.10.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.2
matplotlib       : 3.1.2
numexpr          : 2.7.0
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.2
sqlalchemy       : None
tables           : 3.6.1
xarray           : 0.14.1
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by roberthdevries**
First try to get pandas style tables, if there are none, return the list of non-pandas (hdf5 native) tables in the file (if any)

- [x] closes #29916
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/io/pytables.py
Comment (st-bender) [inline pandas/io/pytables.py]: Hi,
Thanks very much for putting it together!
Why not make "tables" the default and worry about the performance later? (Some "tables"/"table" mixup in the docs btw.)
`dask` for example would have to use the keyword to work properly, which would make it incompatible with
any `pandas < 1.0.x` that doesn't have this keyword.
Given that, I am not so sure anymore to solve that here and in this way, it is an API change that will most certainly break "userspace", but I'd leave that to the maintainers.

File: pandas/tests/io/pytables/test_store.py
Comment (st-bender) [inline pandas/tests/io/pytables/test_store.py]: Shouldn't that be
```python
assert set(store.keys()) == set()
```
?

Comment (roberthdevries) [reply pandas/tests/io/pytables/test_store.py]: This form is equivalent, as the store implements the `__iter__` method which just returns an iterator with `keys()`
I will change it to be explicit.

File: pandas/io/pytables.py
Comment (roberthdevries) [reply pandas/io/pytables.py]: Making "tables" the default is probably breaking the API, but I can do some experiments by changing it and running the unit tests and see what breaks :smile:

Comment (roberthdevries) [reply pandas/io/pytables.py]: Another suggestion that was previously mentioned is to fall back to the "tables" option if the "pandas" option returns no results. This assumes that we either use an HDF5 file that was produced by pandas and only has pandas style tables, or that we read in an HDF5 file that was produced by another application and has HDF5 style tables.
I think that in this way we stay backwards compatible *and* we do not have the performance issue with the pandas style tables.

Comment (st-bender) [reply pandas/io/pytables.py]: Hi,
Again, thanks for taking it on and coming up with a solution.
Looks good to me, so we'll see.
Chhers.

File: pandas/tests/io/pytables/test_store.py
Comment (st-bender) [reply pandas/tests/io/pytables/test_store.py]: Thanks for the explanation, that makes sense indeed.

File: pandas/io/pytables.py
Comment (jreback) [inline pandas/io/pytables.py]: this is not a good api.

either:
a) always return all tables (api breaking), or 
b) add a keyword include='pandas'|'native'|'all', default would be 'pandas'

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/pytables.py
Comment (st-bender) [reply pandas/io/pytables.py]: Note I am not the PR author but reported the issue.

It really depends on what you want to achieve, IMHO better compatibility would be achieved with option (a) which basically reverts #21543 but for `keys()` and leaving `groups()` untouched. Don't know if that breaks anything else, but it might hurt performance in some cases although the original bug  #21372 complained about the performance of `groups()` (it links to #17593 about slow `keys()`). There is also a `walk()` method which is very similar to `groups()`.

(b) would break backwards compatibility of all (future) projects that use that keyword wrt `pandas < 1.0.x`.

Comment (roberthdevries) [reply pandas/io/pytables.py]: The latest version is trying to keep some backwards compatibility with pandas 0.23 (plus or minus a minor revision). Apparently there are some tools that depend on the original behavior (hence the issue which this PR is trying to fix). So either we keep breaking backward compatibility or we add a keyword.
I think this is a matter of taste/policy.

Comment (roberthdevries) [reply pandas/io/pytables.py]: @jreback I think this is a trade-off between having backwards compatible behavior with v0.23.x vs. a clean API.
I see one vote from @st-bender  for being backwards compatible, and I see one vote from @jreback for the clean API
Note sure whose vote has more weight (:smile: but I can make a gues), but I have no strong feelings either way.

Comment (st-bender) [reply pandas/io/pytables.py]: Thanks @roberthdevries for pushing this forward and @jreback thanks for the feedback.
As I see it there is no fully compatible way to change that behaviour back to <=0.23.x, and apparently I am the only one that noticed that when `dask.dataframe.read_hdf()` suddenly failed to read some files which worked before (dask/dask#5934).

It seems that `dask.dataframe` is currently the only serious "user" of that interface, and I filed dask/dask#5934 to decide whether it is better handled within `dask` instead of `pandas`. The change there would be similarly simple and would probably be more backwards compatible with older pandas versions.

**Comment by jreback** [comment]
is this better handled upstream in dask? I am ok with a keyword to .keys but cannot change the behavior like this.

**Comment by roberthdevries** [comment]
I have created a new PR #32723 which has the extra 'include' keyword.
I haven't added the 'all' value as mixing pandas tables and native tables in one file is an anti-pattern (@jreback 's words), which should not be encouraged.

**Comment by st-bender** [comment]
> is this better handled upstream in dask?

I'll try to find out, there was not a lot of response so far, so I'll ping and ask.

> I am ok with a keyword to .keys but cannot change the behavior like this.

I understand and I am with you wrt not changing the API behavoiur.

File: pandas/io/pytables.py
Comment (jamesbooker) [reply pandas/io/pytables.py]: I've just come across this as it's stopping me from working with HDF5 files not produced by pandas.

My issue - how will this cope with files that were produced in some upstream system conforming to the HDF5 specification, then data added to it with pandas (so you have a mix)

Surely the real fix is for pandas to conform to the HDF5 specification (so files work with h5py)

Comment (jreback) [reply pandas/io/pytables.py]: pandas has a specific format of meta-data on top of HDF5. This is not likely to change.

**Comment by jreback** [comment]
I am not sure it is worthile to change anything here. The fact that HDFStore doesn't recognize non-pandas tables is a non-goal here.

**Comment by jamesbooker** [comment]
> I am not sure it is worthile to change anything here. The fact that HDFStore doesn't recognize non-pandas tables is a non-goal here.

This is a disappointing attitude, because this used to work. It's not an enhancement, but a regression that was broken by an 'optimisation'

Groups and keys aren't a feature of pandas, but a feature of the HDF5 standard. If pandas is going to claim support for the HDF5 file format, then it should support the whole format. This isn't even restricted to the dataframe/tables level. You're basically stating the equivalent of 'supporting files stored in a directory not created by pandas is a non-goal'. 

At the dataframe level, I don't understand why you'd not want the ability to load data from a non-pandas table and create a dataframe from it - pandas isn't the only library out there for working with HDF5, and it's safe to assume not all users will be working with data that came out of pandas in the first place (got my hand up here.)

I do, however understand why you'd not want to be converting back from a dataframe to a 'non-pandas table' - being able to read data from a non-pandas HDF5 but only write pandas tables makes some kind of sense.

If pandas is insistent on following this line of thought, then maybe you could at least update the documentation to declare that HDF5 files written not written with pandas won't work with pandas (or declare a derivative file format, called PHDF5 or something and very specifically define what's supported and what's not)

Just one person's opinion.

**Comment by st-bender** [comment]
> > I am not sure it is worthile to change anything here. The fact that HDFStore doesn't recognize non-pandas tables is a non-goal here.
> 
> This is a disappointing attitude, because this used to work. It's not an enhancement, but a regression that was broken by an 'optimisation'
> 
> Groups and keys aren't a feature of pandas, but a feature of the HDF5 standard. If pandas is going to claim support for the HDF5 file format, then it should support the whole format. This isn't even restricted to the dataframe/tables level. You're basically stating the equivalent of 'supporting files stored in a directory not created by pandas is a non-goal'.
> 
> At the dataframe level, I don't understand why you'd not want the ability to load data from a non-pandas table and create a dataframe from it - pandas isn't the only library out there for working with HDF5, and it's safe to assume not all users will be working with data that came out of pandas in the first place (got my hand up here.)

Sorry to jump in here, but **importing** non-pandas files should work just fine and was not the issue here. It is just not possible to list all the keys/groups with standard pandas anymore which lead to downstream problems, most notable with dask. You can have a look at the PR here or at the linked dask issue on how to list all of them with pytables.

If you have trouble importing hdf5 files in general, then that may be a separate issue.

[...]
> If pandas is insistent on following this line of thought, then maybe you could at least update the documentation to declare that HDF5 files written not written with pandas won't work with pandas (or declare a derivative file format, called PHDF5 or something and very specifically define what's supported and what's not)

Yeah I cannot understand that either, but it is the developers' decision to cripple hdf5 support to pandas specific files.

> Just one person's opinion.
> 
> EDIT: I've just noticed that the docs for read_hdf does say 'retrieve a **pandas** object stored in a file' so my comment about the docs above is only half-relevant

That would be a really bad decision indeed.

**Comment by jamesbooker** [comment]
> Sorry to jump in here, 

No need to apologise. It is a discussion, after all

> but **importing** non-pandas files should work just fine and was not the issue here. It is just not possible to list all the keys/groups with standard pandas anymore which lead to downstream problems, most notable with dask. You can have a look at the PR here or at the linked dask issue on how to list all of them with pytables.

The problem isn't just with dask (I'm not using dask). I am trying to import data from non-pandas HDF5 files, to be put into a dataframe, which would then be exported to pandas HDF5 files. My issue was with not being able to list the keys/groups - I have no control of the format of the data coming in, and I need to 'walk' the groups to figure out what's in the file (the file format has a versioning system in it, so I need to walk the groups to figure out how many there are, and where to pull my specific dataset from) - so I guess you could say I have the same issue as dask. They've implemented a workaround which doesn't use the public API of pandas (by drilling down directly into pytables)

> 
> If you have trouble importing hdf5 files in general, then that may be a separate issue.
> 

No, importing the data from a pandas file works fine, my issue is with groups/keys also.

> [...]
> 
> > If pandas is insistent on following this line of thought, then maybe you could at least update the documentation to declare that HDF5 files written not written with pandas won't work with pandas (or declare a derivative file format, called PHDF5 or something and very specifically define what's supported and what's not)
> 
> Yeah I cannot understand that either, but it is the developers' decision to cripple hdf5 support to pandas specific files.

The biggest issue I have is the 'we are not adding new methods' statement. I understand pandas wants a clean API, and certainly doesn't want to roll back a change made so long ago. What I can't fathom is why pandas couldn't have a seperate raw_keys() and raw_groups() method in addition to the keys() and groups() methods. This wouldn't break the existing API, and would allow us to still walk these groups while still keeping in place this 'optimisation'

> 
> > Just one person's opinion.
> > EDIT: I've just noticed that the docs for read_hdf does say 'retrieve a **pandas** object stored in a file' so my comment about the docs above is only half-relevant
> 
> That would be a really bad decision indeed.

What would be a bad decision? I don't understand this last statement.

Thanks for your insight

**Comment by jreback** [comment]
ok I think I commented else where that If you want to support `.keys(nodes='pandas|all')` I think would be ok (and if .groups needs this keyword then ok too).

**Comment by jreback** [comment]
prefer  soln in #32723

**Pull request closed without merging**
